### ISSUE TITLE
How to visualize the CAM/ feature maps?

### ISSUE THREAD

**Issue by cheriku** [open-mmlab/mmsegmentation#2434]
I trained a simple 2 classes fpn-resnet segmentation model based on [configs/sem_fpn/fpn_r50_512x512_160k_ade20k.py](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/sem_fpn/fpn_r50_512x512_160k_ade20k.py). 
Now I hope to visualize the  feature maps, like in vis_cam.py in misclassification or https://github.com/open-mmlab/mmyolo/blob/dev/demo/boxam_vis_demo.py in mmyolo.
I also found and read the following issue, but still failed and confused about how to get the feature maps.
https://github.com/open-mmlab/mmsegmentation/issues/2032#issue-1363956754
Hope someone can provide a more detailed explanation!

**Comment by shenxiangkei** [comment]
Have you found a suitable solution, I have a similar problem. thanks

**Comment by xiexinch** [comment]
```python
import torch
import cv2
import numpy as np

from mmseg.visualization import SegLocalVisualizer
from mmseg.apis import init_model
from mmseg.utils import register_all_modules
from mmengine.model import revert_sync_batchnorm

config_path = '../mmsegv2/configs/pspnet/pspnet_r50-d8_4xb2-40k_cityscapes-512x1024.py'
checkpoint_path = '../mmsegv2/checkpoints/pspnet_r50-d8_512x1024_80k_cityscapes_20200606_112131-2376f12b.pth'
img_path = '../mmsegv2/demo/demo.png'

register_all_modules()

model = init_model(config_path, checkpoint_path, device='cpu')
model = revert_sync_batchnorm(model)
vis = SegLocalVisualizer()


ori_img = cv2.imread(img_path)
img = torch.from_numpy(ori_img.astype(np.single)).permute(2, 0, 1).unsqueeze(0)

logits = model(img)
out = vis.draw_featmap(logits[0], ori_img)

cv2.imshow('cam', out)
cv2.waitKey(0)

```

There might be an example for visualizing the feature maps.

You might refer to https://github.com/open-mmlab/mmengine/blob/cfb884c180e97454b14b1dc12bac21bd376f82db/mmengine/visualization/visualizer.py#L119-L124 for more detail.

**Comment by Yangyang-luo702** [comment]
Thank you very much!
Hope you are happy every day~
Can I add your QQ or Wechat to ask you some questions?
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "open-mmlab/mmsegmentation"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月23日(星期四) 下午4:51
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [open-mmlab/mmsegmentation] How to visualize the CAM/ feature maps? (Issue #2434)





 import torch import cv2 import numpy as np from mmseg.visualization import SegLocalVisualizer from mmseg.apis import init_model from mmseg.utils import register_all_modules from mmengine.model import revert_sync_batchnorm config_path = '../mmsegv2/configs/pspnet/pspnet_r50-d8_4xb2-40k_cityscapes-512x1024.py' checkpoint_path = '../mmsegv2/checkpoints/pspnet_r50-d8_512x1024_80k_cityscapes_20200606_112131-2376f12b.pth' img_path = '../mmsegv2/demo/demo.png' register_all_modules() model = init_model(config_path, checkpoint_path, device='cpu') model = revert_sync_batchnorm(model) vis = SegLocalVisualizer() ori_img = cv2.imread(img_path) img = torch.from_numpy(ori_img.astype(np.single)).permute(2, 0, 1).unsqueeze(0) logits = model(img) out = vis.draw_featmap(logits[0], ori_img) cv2.imshow('cam', out) cv2.waitKey(0)
 
There might be an example for visualizing the feature maps.
 
You might refer to https://github.com/open-mmlab/mmengine/blob/cfb884c180e97454b14b1dc12bac21bd376f82db/mmengine/visualization/visualizer.py#L119-L124 for more detail.

**Comment by Yangyang-luo702** [comment]
![image](https://user-images.githubusercontent.com/82427526/220864190-de501215-1ea5-4b59-83c9-ff72aaa376b3.png)
The package import error is shown here, and the functions are not found in mmsegmentation

**Comment by xiexinch** [comment]
> ![image](https://user-images.githubusercontent.com/82427526/220864190-de501215-1ea5-4b59-83c9-ff72aaa376b3.png) The package import error is shown here, and the functions are not found in mmsegmentation

Could you provide your running env? The version of mmseg? 
These codes are based on mmseg v1.0.0rc5 and mmengine v0.5.0

**Comment by Yangyang-luo702** [comment]
I'm using an old version of mmsegmentation, I'll try version 1.0.0, thanks!



------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2023年2月23日(星期四) 下午5:14
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [open-mmlab/mmsegmentation] How to visualize the CAM/ feature maps? (Issue #2434)





  
 The package import error is shown here, and the functions are not found in mmsegmentation
  
Could you provide your running env? The version of mmseg?
 These codes are based on mmseg v1.0.0rc5 and mmengine v0.5.0

**Comment by xiexinch** [comment]
Closing the issue, as there is no activity for a while.
We hope your issue has been resolved. 
If not, please feel free to open a new one.

**Comment by Yangyang-luo702** [comment]
> Closing the issue, as there is no activity for a while. We hope your issue has been resolved. If not, please feel free to open a new one.

In a semantic partitioned heat map, it is common to draw a heat map for a category, but the function to draw a heat map in mmengine does not seem to have a parameter to select a category.What should I do?

**Comment by graduate0629** [comment]
> ![image](https://user-images.githubusercontent.com/82427526/220864190-de501215-1ea5-4b59-83c9-ff72aaa376b3.png)此处显示了包导入错误，并且在毫米分段中找不到函数

Excuse me, have you solved this problem? How to use it in mmsegmentation 0.30.0?

**Comment by Yangyang-luo702** [comment]
你需要使用mmswgmentation 1.x分支才可以




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "open-mmlab/mmsegmentation"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年8月19日(星期六) 晚上9:53
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [open-mmlab/mmsegmentation] How to visualize the CAM/ feature maps? (Issue #2434)





  
此处显示了包导入错误，并且在毫米分段中找不到函数
  
Excuse me, have you solved this problem? How to use it in mmsegmentation 0.30.0?
